paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/FlinkCatalog.java

```diff
@@ -222,52 +234,106 @@ public void createTable(ObjectPath tablePath, CatalogBaseTable table, boolean ig
+            String value = setOption.getValue();
+
+            if (Objects.equals(oldProperties.get(key), value)) {
+                return null;
+            }
+
+            if (PATH.key().equalsIgnoreCase(key)) {
+                throw new IllegalArgumentException("Illegal table path in table options: " + value);
+            }
+
+            return SchemaChange.setOption(key, value);
+        } else if (change instanceof ResetOption) {
+            ResetOption resetOption = (ResetOption) change;
+            return SchemaChange.removeOption(resetOption.getKey());
+        } else {
+            throw new UnsupportedOperationException(
+                    "Change is not supported: " + change.getClass());
+        }
+    }
+
     @Override
     public void alterTable(
             ObjectPath tablePath, CatalogBaseTable newTable, boolean ignoreIfNotExists)
             throws TableNotExistException, CatalogException {
+        this.alterTable(tablePath, newTable, null, ignoreIfNotExists);
+    }
+
+    @Override
+    public void alterTable(
+            ObjectPath tablePath,
+            CatalogBaseTable newTable,
+            List<TableChange> tableChanges,
+            boolean ignoreIfNotExists)
+            throws TableNotExistException, CatalogException {
         if (ignoreIfNotExists && !tableExists(tablePath)) {
             return;
         }
 
         CatalogTable table = getTable(tablePath);
 
-        // Currently, Flink SQL only support altering table properties.
         validateAlterTable(table, (CatalogTable) newTable);
 
-        List<SchemaChange> changes = new ArrayList<>();
         Map<String, String> oldProperties = table.getOptions();
-        for (Map.Entry<String, String> entry : newTable.getOptions().entrySet()) {
-            String key = entry.getKey();
-            String value = entry.getValue();
-
-            if (Objects.equals(value, oldProperties.get(key))) {
-                continue;
-            }
-
-            if (PATH.key().equalsIgnoreCase(key)) {
-                throw new IllegalArgumentException("Illegal table path in table options: " + value);
-            }
-
-            changes.add(SchemaChange.setOption(key, value));
+        List<SchemaChange> changes = new ArrayList<>();
+        if (null != tableChanges) {
+            List<SchemaChange> schemaChanges =
+                    tableChanges.stream()
+                            .map(m -> toSchemaChange(oldProperties, m))
+                            .filter(Objects::nonNull)
+                            .collect(Collectors.toList());
+            changes.addAll(schemaChanges);
         }
 
-        oldProperties
-                .keySet()
-                .forEach(
-                        k -> {
-                            if (!newTable.getOptions().containsKey(k)) {
-                                changes.add(SchemaChange.removeOption(k));
-                            }
-                        });
-
```

Comment:
> We cannot delete this because < 1.17 version still use the old style `alterTable` API. How about don't make `alterTable(ObjectPath tablePath, CatalogBaseTable newTable, boolean ignoreIfNotExists)` call the new `alterTable`? To verify that old `alter table set option` still works, we can copy old tests in `ShemaChangeITCase` to Flink 1.16 module.

I copy the previous `ShemaChangeITCase ` and `FlinkCatalog` to flink 1.14, 1.15, 1.16 , in  flink 1.17 ，I think it's best to unify the logic, use  `FlinkCatalog#toSchemaChange` to get the change, what do you think about ?

Reply:
I left a comment at the bottom about this.